File: ts/sdk/src/geometry/vec3.ts
/** Immutable three-component vector in the caller's coordinate frame. */
export class Vec3 {
  constructor(
    public readonly x: number,
    public readonly y: number,
    public readonly z: number,
  ) {}

  add(other: Vec3): Vec3 { return new Vec3(this.x + other.x, this.y + other.y, this.z + other.z) }
  sub(other: Vec3): Vec3 { return new Vec3(this.x - other.x, this.y - other.y, this.z - other.z) }
  mul(scale: number): Vec3 { return new Vec3(this.x * scale, this.y * scale, this.z * scale) }
  negated(): Vec3 { return new Vec3(-this.x, -this.y, -this.z) }
  dot(other: Vec3): number { return this.x * other.x + this.y * other.y + this.z * other.z }
  cross(other: Vec3): Vec3 {
    return new Vec3(
      this.y * other.z - this.z * other.y,
      this.z * other.x - this.x * other.z,
      this.x * other.y - this.y * other.x,
    );
  }
  length(): number { return Math.sqrt(this.dot(this)) }
  normalized(): Vec3 {
    const length = this.length();
    return length > 1e-12 ? this.mul(1 / length) : new Vec3(0, 0, 0);
  }
  toArray(): [number, number, number] { return [this.x, this.y, this.z] }
  static fromArray(value: readonly [number, number, number]): Vec3 { return new Vec3(value[0], value[1], value[2]) }
}

File: cs/OpenAxis/Geometry/Vec3.cs
using System;

namespace OpenAxis.Geometry
{
    /// <summary>3D vector with basic linear algebra operations.</summary>
    public readonly struct Vec3
    {
        public readonly double X;
        public readonly double Y;
        public readonly double Z;

        public Vec3(double x, double y, double z) { X = x; Y = y; Z = z; }

        public static Vec3 operator +(Vec3 a, Vec3 b) => new Vec3(a.X + b.X, a.Y + b.Y, a.Z + b.Z);
        public static Vec3 operator -(Vec3 a, Vec3 b) => new Vec3(a.X - b.X, a.Y - b.Y, a.Z - b.Z);
        public static Vec3 operator *(Vec3 v, double s) => new Vec3(v.X * s, v.Y * s, v.Z * s);
        public static Vec3 operator *(double s, Vec3 v) => new Vec3(v.X * s, v.Y * s, v.Z * s);
        public static Vec3 operator -(Vec3 v) => new Vec3(-v.X, -v.Y, -v.Z);

        public double Dot(Vec3 o) => X * o.X + Y * o.Y + Z * o.Z;

        public Vec3 Cross(Vec3 o) => new Vec3(
            Y * o.Z - Z * o.Y,
            Z * o.X - X * o.Z,
            X * o.Y - Y * o.X);

        public double Length() => Math.Sqrt(Dot(this));

        public Vec3 Normalized()
        {
            var n = Length();
            return n > 1e-12 ? new Vec3(X / n, Y / n, Z / n) : new Vec3(0, 0, 0);
        }

        public override string ToString() => $"Vec3({X:G}, {Y:G}, {Z:G})";
    }
}
